### PR TITLE
Document code: Add tree-sitter query for Java

### PR DESCRIPTION
**Pull request by umpox**
image

## Description

This PR adds tree-sitter support for `getDocumentableNode` in Java.

This means we get:
- Improved range expansion
- Code actions appearing automatically against documentable nodes
- Ghost text appearing above documentable nodes (assuming there is no containing documentable node)

## Test plan

1. Open Java code
2. Move cursor to a position on a documentable node, e.g. a method name
3. Document code, expect documentation is generated correctly (above the method)
4. Move cursor to a position within a documentable node, e.g. within a method
5. Document code, expect documentation is generated correctly (above the method)

### CODE REVIEW

File: vscode/src/tree-sitter/query-tests/test-data/documentable-node.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+public enum Day {
+//           |
+    SUNDAY,
+    MONDAY,
+    TUESDAY,
+    WEDNESDAY,
+    THURSDAY,
+    FRIDAY,
+    SATURDAY
```
Comment (umpox) [inline vscode/src/tree-sitter/query-tests/test-data/documentable-node.java:75]: Note: I haven't made enum members documentable here. Happy to add them if anyone thinks that's useful

File: vscode/src/tree-sitter/queries/java.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+`
+
+export const javaQueries = {
+    [SupportedLanguage.java]: {
+        documentableNodes: DOCUMENTABLE_NODES,
+        singlelineTriggers: '',
+        intents: '',
+        identifiers: '',
+        graphContextIdentifiers: '',
+        enclosingFunction: '',
```
Comment (umpox) [inline vscode/src/tree-sitter/queries/java.ts:37]: @valerybugakov We should consider adding autocomplete queries here too

File: vscode/src/tree-sitter/query-tests/test-data/documentable-node.java (hunk 1 above)
Comment (olafurpg) [reply vscode/src/tree-sitter/query-tests/test-data/documentable-node.java:75]: I think it's OK to make them documentable. Java enums are a lot more powerful than TS enums, they are just regular classes and each enum member can be initialized with custom parameters.

**Review by pkukielka — APPROVED** [review]
LGTM

File: vscode/src/tree-sitter/query-tests/test-data/documentable-node.java (hunk 1 above)
Comment (umpox) [reply vscode/src/tree-sitter/query-tests/test-data/documentable-node.java:75]: Oh wow yeah they can be pretty involved... I had to update the query in a few places, added some more tests. Would appreciate your eyes again!

**Comment by olafurpg** [comment]
I tried this locally and "Document Code" is still not going to the enclosing *method declaration*, it's documenting local variables. Any idea what might be going on @umpox ? To reproduce https://sourcegraph.com/github.com/redisson/redisson@68f6829e7ab22c49840e17825e43f944956f9785/-/blob/redisson/src/main/java/org/redisson/cluster/ClusterConnectionManager.java?L894

https://github.com/sourcegraph/cody/assets/1408093/17d66a42-efa1-42d5-8a33-b87b50150d77

**Comment by umpox** [comment]
@olafurpg That's intentional, as it matches the same behaviour for this command in other languages.

E.g. if the user wants to generate docs for a local variable in a long function

image

We do have an `enclosingFunction` query, but that is used for the edit command specifically (where you will rarely want to expand to a local variable). It's not added by this PR but we can add that as a follow up.

**Comment by olafurpg** [comment]
Sounds good. Worth noting that @steveyegge 's Loom showed a demo where documenting a local block would document the enclosing method. We need a bigger discussion on what the desired behavior should be.

This PR LGTM because it makes Java work more similarly to other languages. It would be nice to add more coverage for documentable ranges in languages that see heavier usage in JetBrains, but it can be done separately.

**Comment by umpox** [comment]
@olafurpg Agree, I think @abeatrix is looking at doing the same for Kotlin today. I can follow up on this for `enclosingFunction` and then I'll look into how much effort it'll be to add PHP support too

**Pull request merged**